### PR TITLE
Add BaseVideoProvider::getReferenceUrl() method

### PR DESCRIPTION
**Pull request by core23**
Rebase of https://github.com/sonata-project/SonataMediaBundle/pull/622

### Changelog

```markdown
### Added
- Created `getReferenceUrl` method for all video providers
```

### Subject

Created a new method which returns the reference url of a video.

### To do

- [X] Update the tests

### CODE REVIEW

File: Provider/BaseVideoProvider.php [hunk 1]
```diff
@@ -179,6 +179,16 @@ public function postRemove(MediaInterface $media)
...
 
+    // NEXT_MAJOR: Uncomment this line
+    /*
+     * Get provider reference url.
+     *
+     * @param MediaInterface $media
+     *
+     * @return string
+     */
+    // public function getReferenceUrl(MediaInterface $media);
```
Comment (greg0ire) [inline Provider/BaseVideoProvider.php]: Should be abstract ?

File: Provider/DailyMotionProvider.php [hunk 2]
```diff
@@ -125,7 +125,22 @@ public function updateMetadata(MediaInterface $media, $force = false)
      */
     public function getDownloadResponse(MediaInterface $media, $format, $mode, array $headers = array())
     {
-        return new RedirectResponse(sprintf('http://www.dailymotion.com/video/%s', $media->getProviderReference()), 302, $headers);
+        return new RedirectResponse($this->getReferenceUrl($media), 302, $headers);
+    }
+
+    /**
+     * {@inheritdoc}
```
Comment (OskarStark) [inline Provider/DailyMotionProvider.php]: not true, this can be done in the master PR afterwards

File: Provider/DailyMotionProvider.php [hunk 3]
```diff
@@ -125,7 +125,22 @@ public function updateMetadata(MediaInterface $media, $force = false)
...
+     */
+    public function getReferenceUrl(MediaInterface $media)
+    {
+        return sprintf('http://www.dailymotion.com/video/%s', $media->getProviderReference());
+    }
+
+    public function testGetReferenceUrl()
```
Comment (OskarStark) [inline Provider/DailyMotionProvider.php]: why a test method here? oO

**Comment by OskarStark** [comment]
Test für DailymotionProvider is missing

File: Provider/DailyMotionProvider.php (hunk 3 above)
Comment (Soullivaneuh) [inline Provider/DailyMotionProvider.php]: :wine_glass: 

Or maybe :beers:?

:wink:

**Comment by Soullivaneuh** [comment]
@core23 Could you please add some documentation.

Also, even if it's not a deprecation, I think it would be great to have an upgrade one because this method would be mandatory on next major.

**Comment by core23** [comment]
@Soullivaneuh I don't agree with that. Every change in the interface / abstract class was held in the major change where we finally introduce this method. I would do this in the major PR.

**Comment by core23** [comment]
Ping @Soullivaneuh

File: Tests/Provider/YouTubeProviderTest.php [hunk 4]
```diff
@@ -206,4 +206,11 @@ public function testHelperProperties()
         $this->assertSame(100, $properties['player_parameters']['height']);
         $this->assertSame(100, $properties['player_parameters']['width']);
     }
+
+    public function testGetReferenceUrl()
+    {
+        $media = new Media();
+        $media->setProviderReference('123456');
+        $this->assertEquals('http://www.youtube.com/watch?v=123456', $this->getProvider()->getReferenceUrl($media));
+    }
```
Comment (greg0ire) [inline Tests/Provider/YouTubeProviderTest.php]: This is getting repetitive… maybe you could introduce an abstract test case with a `getExpectedUrl($providerReference)` method ?

**Comment by Soullivaneuh** [comment]
> @Soullivaneuh I don't agree with that. Every change in the interface / abstract class was held in the major change where we finally introduce this method. I would do this in the major PR

Don't know. Ping @sonata-project/contributors for opinions.

**Comment by greg0ire** [comment]
Well one could see not implementing the new method as "deprecated". It would make the transition smoother, that's for sure…

File: Tests/Provider/YouTubeProviderTest.php (hunk 4 above)
Comment (core23) [inline Tests/Provider/YouTubeProviderTest.php]: Note the scope of this PR. We could move more methods to an abstract TestCase.

**Comment by core23** [comment]
> Well one could see not implementing the new method as "deprecated". It would make the transition smoother, that's for sure… this could be done by adding a check in the constructor

Don't thinkt that's a good idea. This would mess up the constructor and slow down the performance or is it the symfony way?

File: Tests/Provider/YouTubeProviderTest.php (hunk 4 above)
Comment (greg0ire) [inline Tests/Provider/YouTubeProviderTest.php]: Ok

**Comment by greg0ire** [comment]
> This would mess up the constructor and slow down the performance or is it the symfony way?

Regarding the performance, if you're thinking about reflection, I think this can be achieved without. Just `method_exists` would suffice IMO

With method exists : https://3v4l.org/dsEdb/perf#tabs
With nothing : https://3v4l.org/bOFuq/perf#tabs

**Comment by core23** [comment]
Added the `deprecation` warning. Don't know what kind of docs are missing.

File: Provider/BaseVideoProvider.php [hunk 5]
```diff
@@ -179,6 +183,16 @@ public function postRemove(MediaInterface $media)
     {
     }
 
+    // NEXT_MAJOR: Uncomment this line
```
Comment (OskarStark) [inline Provider/BaseVideoProvider.php]: `Uncomment this line` -> `Uncomment this method`

File: Provider/DailyMotionProvider.php (hunk 2 above)
Comment (OskarStark) [inline Provider/DailyMotionProvider.php]: or could this inherit from an uncommented method?

File: Provider/BaseVideoProvider.php (hunk 5 above)
Comment (core23) [inline Provider/BaseVideoProvider.php]: Fixed

**Comment by core23** [comment]
@Soullivaneuh Can you explain what documentation are you expecting?

**Comment by core23** [comment]
Ping @Soullivaneuh

**Comment by core23** [comment]
@Soullivaneuh ?

**Comment by Soullivaneuh** [comment]
> @Soullivaneuh Can you explain what documentation are you expecting?

You introduced a new method, right? This one should be documented. :-)

**Comment by core23** [comment]
> You introduced a new method, right? This one should be documented. :-)

Have a look at the phpDoc ;)

**Comment by Soullivaneuh** [comment]
I'm talking about the RST documentation, with some samples. This is the only goal of this label BTW.

**Comment by greg0ire** [comment]
Here is probably where the doc should be added @core23 : https://github.com/sonata-project/SonataMediaBundle/blob/3.x/Resources/doc/reference/creating_a_provider_class.rst

**Comment by core23** [comment]
Added the missing docs.

**Comment by greg0ire** [comment]
LGTM

File: Provider/BaseVideoProvider.php [hunk 6]
```diff
@@ -50,6 +50,10 @@ public function __construct($name, Filesystem $filesystem, CDNInterface $cdn, Ge
     {
         parent::__construct($name, $filesystem, $cdn, $pathGenerator, $thumbnail);
 
+        if (!method_exists($this, 'getReferenceUrl')) {
```
Comment (OskarStark) [inline Provider/BaseVideoProvider.php]: `NEXT_MAJOR: remove this check!`

**Comment by OskarStark** [comment]
After my comment, this PR is `RTM`

**Comment by core23** [comment]
Fixed @OskarStark

**Comment by greg0ire** [comment]
Thanks @core23!

**Pull request merged**